### ISSUE TITLE
Bug: Inline `ref` on Any Media Component Detaches and Re-attaches the Engine on Every Parent Render

### ISSUE THREAD

**Issue by rennanribeiro** [videojs/v10#2725]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Which package(s) or projects are affected?

React (@videojs/react)

### Browser/OS/Node environment

@videojs/react@10.0.0-rc.2 (main at 68bcf5b0), react@19.2.8, hls.js@1.6.15, Chromium (Edge 140 headless and Brave) on macOS 15

## Summary

Every media component composes the ref that attaches the engine with the consumer's forwarded ref through `useComposedRefs(attachRef, ref)`. The hook memoizes on the refs themselves, so the composed callback takes a new identity whenever the forwarded ref does. An inline `ref={(el) => ...}` is a new function every render, which React allows and player UIs do all the time, and React runs the cleanup of a ref callback whose identity changed before calling the new one. The cleanup composed from `useAttachMedia` is `media.detach()`, the new callback runs `media.attach(element)`. For hls.js that is `detachMedia()` (removes `src`, calls `load()`) followed by `attachMedia()`: the MediaSource is torn down and rebuilt on every parent render.

This is the churn from #1296 coming back through the other input. That fix memoized the Provider value and replaced `attachMediaElement`, the internal sources of a fresh callback ref, but the composed callback is still coupled to whatever ref the consumer passes, and the JSDoc still promises "Memoized for stable reference".

## Reproduction

Two `HlsJsVideo` elements, same Mux test stream, `autoPlay muted playsInline preload="auto"`, inside a parent that re-renders every 500 ms (a `setInterval` state tick standing in for a time display). Left: `ref={useCallback((el) => track(el), [])}`. Right: `ref={(el) => track(el)}`. The event counters are bound once per element, so the probe itself never restarts anything.

```tsx
function App() {
  const [, setTick] = useState(0);
  const videoRef = useRef<HTMLVideoElement | null>(null);

  useEffect(() => {
    const id = setInterval(() => setTick((t) => t + 1), 500);
    return () => clearInterval(id);
  }, []);

  return (
    <HlsJsVideo
      ref={(el) => { videoRef.current = el; }}
      src="https://stream.mux.com/VcmKA6aqzIzlg3MayLJDnbF55kX00mds028Z65QxvBYaA.m3u8"
      autoPlay muted playsInline preload="auto"
    />
  );
}
```

After 20 s, one player per page (headless Edge, same stream):

| | stable ref | inline ref |
|---|---|---|
| `currentTime` | 19.80 | 8.62 |
| `emptied` / `loadstart` | 0 / 1 | 39 / 40 |
| `waiting` | 0 | 39 |
| playlists / segments fetched | 2 / 12 | 3 / 467 |

Every parent render logs `emptied, loadstart, waiting, loadedmetadata, playing`. The picture blanks for the reset, playback advances at less than half speed and the CDN gets 39x the segment traffic. Side by side on one page it is the same picture (18.78 s vs 8.26 s after 20 s).

![stable ref and inline ref side by side after 20 s](https://raw.githubusercontent.com/rennanribeiro/v10/issue-assets/composed-ref/01-both-20s.png)

![inline ref alone after 20 s: 39 resets, 467 segments, black frame during a reset](https://raw.githubusercontent.com/rennanribeiro/v10/issue-assets/composed-ref/03-inline-20s.png)

With the fix, same page and stream:

![inline ref with the fix after 20 s: 0 resets, 12 segments](https://raw.githubusercontent.com/rennanribeiro/v10/issue-assets/composed-ref/after/03-inline-20s.png)

## Root cause

`packages/react/src/utils/use-composed-refs.ts`:

```ts
export function useComposedRefs<T>(...refs: OptionalRef<T>[]): RefCallback<T> {
  return useCallback(composeRefs(...refs), [...refs]);
}
```

`composeRefs` returns a cleanup whenever any composed ref returned one, and the `useAttachMedia` ref always does (`() => media.detach?.()`). React 19 treats a ref callback with a new identity as unmount plus mount of the ref: cleanup, then the new callback with the element. React 18 gets to the same place through `ref(null)` then `ref(element)`, and `useAttachMedia` maps `null` to `media.detach()`. Either way, a new identity of any one composed ref releases and rebinds all of them.

Every adapter composes `attachRef` with the forwarded `ref` this way: hlsjs-video, hls-video, hls-audio, native-hls-video, dash-video, shaka-video, mux-video, mux-audio, hls-background-video, and the embeds (youtube, vimeo, twitch, tiktok, wistia, cloudflare, spotify) through `eventsRef, attachRef, ref`.

## Fix

Branch: https://github.com/videojs/v10/compare/main...rennanribeiro:v10:fix/react-composed-ref-identity

The first version of this branch changed `useComposedRefs`; after the review below it was replaced. `useComposedRefs` is now untouched and the change lives where the churn actually comes from:

- `useAttachMedia` remembers what the adapter is attached to and only attaches when the element or the adapter actually changed, so the `null` then same-element pair React hands a composed ref that changed identity is a no-op. Attaching still happens while the refs run, ahead of any effect, exactly where it did before. A layout effect that runs after the refs of each commit detaches when the element went away with nothing to follow (and reattaches after a StrictMode simulated unmount); another detaches on unmount.
- `useAttachIframe`, a copy of the old hook used by the six embed adapters (YouTube, Vimeo, Twitch, TikTok, Cloudflare, Spotify), now delegates to it. There the churn was worse: a re-attach destroys and rebuilds the embed player.
- Reference page sentence updated to the new contract.
- Tests: nine cases in `use-attach-media.test.tsx` (mount/unmount, identity churn, attach before the mounting commit's effects, listeners composed ahead run first, element swap, adapter swap with detach-before-attach order, element removed while mounted, adapter that fails to attach, StrictMode) plus `media/tests/hlsjs-video.test.tsx` against the real `HlsJsAdapter`. Four fail on main; the `@videojs/react` suite is green (583 tests), lint, typecheck and check:workspace pass. Re-measured on the same page: inline ref now plays 19.3 s in 20 s, 0 resets, 12 segments.

Related, not in this branch: `renderElement` -> `mergeRefs` -> `composeRefs` still builds a fresh callback every render for the 36 `ref: [forwardedRef, xRef]` call sites (VolumeSlider's `wheelRef` re-adds its listener on every render), and `HlsJsAdapter.attach` forwards a repeat attach of the same element to `hls.attachMedia`, which hls.js answers with a detach and re-attach. Happy to take both next.

Happy to open this as a PR if that can be enabled, or for a collaborator to pick the branch up.

**Comment by luwes** [comment]
thanks for reporting this bug! I quickly scanned the potential fix and it feels off to me that such a big change is needed to something like `useComposedRefs` which should be a generic utility. I believe we adopted this from another project. 

Can this be done in another way keeping the utility intact?

**Comment by rennanribeiro** [comment]
You're right, and thanks for the push back. The hook is Radix's and should stay that way; the churn comes from `useAttachMedia` attaching inside the ref callback, so that is where the fix lives now. Branch replaced, same link: `useAttachMedia` remembers what the adapter is attached to and only attaches when the element or the adapter actually changed, so the `null` then same-element pair React hands a composed ref that changed identity is a no-op. Attaching still happens during the ref phase, exactly where it did before; a layout effect covers unmount and an element that goes away. `useAttachIframe`, the copy the six embed adapters use (where a re-attach rebuilds the embed player), now delegates to it. `useComposedRefs` is untouched.

10 tests, 4 fail on main, react suite green, and the same page re-measured: the inline ref now plays 19.3 s in 20 s with 0 resets and 12 segments. Details in the Fix section above.

One note for whoever reviews: I first tried recording the element as state and attaching in an effect, the `Video`/`setMedia` shape. It also removes the churn, but it moves the attach to a second commit, after the mounting commit's effects, so a `useEffect(() => ref.current.play())` on a media component would call play() before hls.js sets `src`. Keeping the attach in the ref phase avoids that.
